### PR TITLE
[SYCL][BLAS] Move portBLAS to UXL Foundation as generic SYCL BLAS

### PR DESCRIPTION
**Pull request by s-Nick**
# Description

This PR aims to keep oneMath portBLAS backend compatible and usable after the upcoming changes to it. Everything will be moved to UXL Foundation as a generic SYCL BLAS implementation. 
Here all occurences to portBLAS are removed and renamed to a more appropriate name.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
[intel_igpu_log.txt](https://github.com/user-attachments/files/18092542/intel_igpu_log.txt)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ oneMath is part of the [UXL Foundation](http://www.uxlfoundation.org).
             <td align="center">AMD GPU</td>
         </tr>
         <tr>
-            <td align="center"><a href="https://github.com/codeplaysoftware/portBLAS"> portBLAS </a></td>
+            <td align="center"><a href="https://github.com/codeplaysoftware/portBLAS"> generic SYCL BLAS </a></td>
```
Comment (s-Nick) [inline README.md]: There many links that still mention and bring to portBLAS codebase. I will update them as soon as the migration is completed.

File: docs/building_the_project_with_dpcpp.rst [hunk 2]
```diff
@@ -236,18 +236,18 @@ SYCL enables portable heterogeneous computing on a wide range of accelerators.
...
+any other option required for performance. See `Building for oneMath generic SYCL BLAS`_ and
 `Building for portFFT`_. Extensive testing is strongly advised for these
 unsupported configurations.
 
 .. _build_for_portlibs_dpcpp:
 
-Pure SYCL backends: portBLAS and portFFT
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Pure SYCL backends: generic BLAS and portFFT
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (Rbiessy) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It's missing 2 to underline the whole title

File: docs/building_the_project_with_dpcpp.rst [hunk 3]
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
...
-Both portBLAS and portFFT are used as header-only libraries, and will be
+Both onemath_sycl_blas and portFFT are used as header-only libraries, and will be
 downloaded automatically if not found.
 
-.. _build_for_portblas_dpcpp:
+.. _build_for_generic_blas_dpcpp:
 
-Building for portBLAS
+Building for oneMath generic SYCL BLAS
 ---------------------
```
Comment (Rbiessy) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
Building for oneMath generic SYCL BLAS
--------------------------------------
```

File: docs/building_the_project_with_dpcpp.rst [hunk 4]
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
...
-`portBLAS <https://github.com/codeplaysoftware/portBLAS>`_ is
-enabled by setting ``-DENABLE_PORTBLAS_BACKEND=True``.
+`onemath_sycl_blas generic BLAS <https://github.com/codeplaysoftware/portBLAS>`_ is
+enabled by setting ``-DENABLE_GENERIC_BLAS_BACKEND=True``.
 
-By default, the portBLAS backend is not tuned for any specific device.
+By default, the generic BLAS backend is not tuned for any specific device.
 This tuning is required to achieve best performance.
-portBLAS can be tuned for a specific hardware target by adding compiler
+OneMath_sycl_blas can be tuned for a specific hardware target by adding compiler
```
Comment (Rbiessy) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
OneMath SYCL BLAS can be tuned for a specific hardware target by adding compiler
```
I think this is more consistent with the rest of the document?

File: docs/building_the_project_with_dpcpp.rst [hunk 5]
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
 For best performance, both libraries must be tuned. See the individual sections
 for more details.
 
-Both portBLAS and portFFT are used as header-only libraries, and will be
+Both onemath_sycl_blas and portFFT are used as header-only libraries, and will be
```
Comment (Rbiessy) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
Both oneMath SYCL BLAS and portFFT are used as header-only libraries, and will be
```

File: docs/building_the_project_with_dpcpp.rst [hunk 6]
```diff
@@ -59,7 +59,7 @@ or ``clang++`` and ``clang`` respectively when using the Open DPC++ Compiler.
 Backends should be enabled by setting ``-DENABLE_<BACKEND_NAME>_BACKEND=True``
 for each desired backend. By default, only the ``MKLGPU`` and ``MKLCPU``
 backends are enabled. Multiple backends for multiple device vendors can be
-enabled at once (albeit with limitations when using portBLAS and portFFT). The
+enabled at once (albeit with limitations when using onemath_sycl_blas and portFFT). The
```
Comment (Rbiessy) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
enabled at once (albeit with limitations when using oneMath SYCL BLAS and portFFT). The
```

File: docs/building_the_project_with_dpcpp.rst [hunk 7]
```diff
@@ -59,7 +59,7 @@ or ``clang++`` and ``clang`` respectively when using the Open DPC++ Compiler.
 Backends should be enabled by setting ``-DENABLE_<BACKEND_NAME>_BACKEND=True``
 for each desired backend. By default, only the ``MKLGPU`` and ``MKLCPU``
 backends are enabled. Multiple backends for multiple device vendors can be
-enabled at once (albeit with limitations when using portBLAS and portFFT). The
+enabled at once (albeit with limitations when using oneMath SYCL BLAS and portFFT). The
```
Comment (andrewtbarker) [inline docs/building_the_project_with_dpcpp.rst]: Should this be "generic SYCL BLAS" or "oneMath generic SYCL BLAS"?

File: docs/building_the_project_with_dpcpp.rst [hunk 8]
```diff
@@ -236,18 +236,18 @@ SYCL enables portable heterogeneous computing on a wide range of accelerators.
...
 
 .. _build_for_portlibs_dpcpp:
 
-Pure SYCL backends: portBLAS and portFFT
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Pure SYCL backends: generic BLAS and portFFT
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-`portBLAS <https://github.com/codeplaysoftware/portBLAS>`_ and `portFFT
+`Generic BLAS <https://github.com/codeplaysoftware/portBLAS>`_ and `portFFT
```
Comment (andrewtbarker) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
`Generic SYCL BLAS <https://github.com/codeplaysoftware/portBLAS>`_ and `portFFT
```

File: docs/building_the_project_with_dpcpp.rst [hunk 9]
```diff
@@ -435,11 +435,11 @@ Build oneMath for the BLAS domain on a generic SYCL device:
...
       -DENABLE_MKLCPU_BACKEND=False \ 
       -DENABLE_MKLGPU_BACKEND=False \
-      -DENABLE_PORTBLAS_BACKEND=True
+      -DENABLE_GENERIC_BLAS_BACKEND=True
 
 Note that this is not a tested configuration. This builds oneMath with the
-portBLAS backend only, for a generic SYCL device supported by the Open DPC++
-project.
+generic BLAS backend only, for a generic SYCL device supported by the
+Open DPC++ project.
```
Comment (andrewtbarker) [inline docs/building_the_project_with_dpcpp.rst]: What is the "Open DPC++ project"? I recommend ending the sentence after "for a generic SYCL device."

File: docs/building_the_project_with_dpcpp.rst [hunk 10]
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
...
-`portBLAS <https://github.com/codeplaysoftware/portBLAS>`_ is
-enabled by setting ``-DENABLE_PORTBLAS_BACKEND=True``.
+`onemath_sycl_blas generic BLAS <https://github.com/codeplaysoftware/portBLAS>`_ is
+enabled by setting ``-DENABLE_GENERIC_BLAS_BACKEND=True``.
 
-By default, the portBLAS backend is not tuned for any specific device.
+By default, the generic BLAS backend is not tuned for any specific device.
 This tuning is required to achieve best performance.
-portBLAS can be tuned for a specific hardware target by adding compiler
+OneMath SYCL BLAS can be tuned for a specific hardware target by adding compiler
```
Comment (andrewtbarker) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
The generic BLAS backend can be tuned for a specific hardware target by adding compiler
```

File: docs/building_the_project_with_dpcpp.rst [hunk 11]
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
 For best performance, both libraries must be tuned. See the individual sections
 for more details.
 
-Both portBLAS and portFFT are used as header-only libraries, and will be
+Both oneMath SYCL BLAS and portFFT are used as header-only libraries, and will be
```
Comment (andrewtbarker) [inline docs/building_the_project_with_dpcpp.rst]:
```suggestion
Both generic SYCL BLAS and portFFT are used as header-only libraries, and will be
```

**Review by andrewtbarker — COMMENTED** [review]
Generally looks good, I think the use of names in the documentation could be a bit more consistent.

One question (which probably should have been brought up earlier in this process, sorry): "generic BLAS" to me suggests any library that implements the BLAS standard (eg ESSL, AOCL, NVPL, BLIS, ...) but that is not what is meant here. So I guess the suggestion is to include "SYCL" a lot for clarity.

**Comment by s-Nick** [comment]
Thank you for your review @andrewtbarker 
I understand what you mean, I'll give another look to documentation for name consistency. In the mean time, I addressed all your suggestions in [b88b452](https://github.com/uxlfoundation/oneMath/pull/618/commits/b88b4527da5a32fe7a6529ebbb94704a12b1a7a1)

File: README.md (hunk 1 above)
Comment (s-Nick) [reply README.md]: Updated in [2abd2c7](https://github.com/uxlfoundation/oneMath/pull/618/commits/2abd2c7eb9ba7a90e0f93c8166c3df8fe5652ec4)

**Review by andrewtbarker — APPROVED** [review]

**Review by Rbiessy — APPROVED** [review]
Thanks LGTM, if there are no more concerns I will merge this tomorrow morning!

File: CMakeLists.txt [hunk 12]
```diff
@@ -45,7 +45,7 @@ endif()
 option(ENABLE_CUBLAS_BACKEND "Enable the cuBLAS backend for the BLAS interface" OFF)
 option(ENABLE_ROCBLAS_BACKEND "Enable the rocBLAS backend for the BLAS interface" OFF)
 option(ENABLE_NETLIB_BACKEND "Enable the Netlib backend for the BLAS interface" OFF)
-option(ENABLE_PORTBLAS_BACKEND "Enable the portBLAS backend for the BLAS interface. Cannot be used with other BLAS backends." OFF)
+option(ENABLE_GENERIC_BLAS_BACKEND "Enable the generic BLAS backend for the BLAS interface. Cannot be used with other BLAS backends." OFF)
```
Comment (Rbiessy) [inline CMakeLists.txt:48]: Can you add some logic below to check if the user defines `ENABLE_PORTBLAS_BACKEND`? I suggest that we add a warning message explaining is `ENABLE_PORTBLAS_BACKEND` deprecated. We can automatically set `ENABLE_GENERIC_BLAS_BACKEND=ON` in this case to keep the internal CI working.

File: src/blas/backends/generic/CMakeLists.txt [hunk 13]
```diff
@@ -17,11 +17,11 @@
 #
 #=========================================================================
 
-set(LIB_NAME onemath_blas_portblas)
+set(LIB_NAME onemath_blas_generic)
 set(LIB_OBJ ${LIB_NAME}_obj)
 
-if(NOT DEFINED PORTBLAS_TUNING_TARGET)
```
Comment (Rbiessy) [inline src/blas/backends/generic/CMakeLists.txt:23]: Similarly to my comment, can we warn the user but keep things working if `PORTBLAS_TUNING_TARGET` is set?

**Review by Rbiessy — CHANGES_REQUESTED** [review]
I noticed the GitHub CI is not running the portBLAS tests anymore as it still uses the portBLAS options. Can you rename the occurrences in https://github.com/uxlfoundation/oneMath/blob/develop/.github/workflows/pr.yml#L37?

File: src/blas/backends/generic/CMakeLists.txt (hunk 13 above)
Comment (s-Nick) [reply src/blas/backends/generic/CMakeLists.txt:23]: I added a warning message for both in  [e8d22aa](https://github.com/uxlfoundation/oneMath/pull/618/commits/e8d22aac5b8e6cd0d2d7e69aa0e1412eb193b79c)

**Comment by s-Nick** [comment]
@Rbiessy I updated the workflows in [a97bcb2](https://github.com/uxlfoundation/oneMath/pull/618/commits/a97bcb2d686391bb14c58f0b05b5412591af2ece) and it is running the CI now

**Review by Rbiessy — APPROVED** [review]

**Comment by zettai-reido** [comment]
Looks approachable.
Do we really need those pre-C++17 namespaces in cpp files instead of
`onemath::blas::generic` style?

**Comment by s-Nick** [comment]
Thank you for your feedback @zettai-reido.
There is not particular reason for that namespace a part from following the contributing guideline (NS4) https://github.com/uxlfoundation/oneMath/blob/develop/CONTRIBUTING.md#ns-namespaces
I think this change is outside the scope of this PR. If we  want to update the style we would need to update the documentation and other backend. Could you open an issue for it?

**Review by andreyfe1 — APPROVED** [review]
Looks good to me overall. Thanks!

**Pull request merged**